src/lib/shutdown/shutdown.cpp

```diff
@@ -61,94 +68,139 @@
 
 using namespace time_literals;
 
-static pthread_mutex_t shutdown_mutex =
-	PTHREAD_MUTEX_INITIALIZER; // protects access to shutdown_hooks & shutdown_lock_counter
-static uint8_t shutdown_lock_counter = 0;
+static pthread_mutex_t shutdown_hook_mutex =
+	PTHREAD_MUTEX_INITIALIZER; // protects access hook registeration
 
-int px4_shutdown_lock()
-{
-	int ret = pthread_mutex_lock(&shutdown_mutex);
+#if defined(CONFIG_SCHED_WORKQUEUE) || (!defined(CONFIG_BUILD_FLAT) && defined(CONFIG_LIBC_USRWORK))
 
-	if (ret == 0) {
-		++shutdown_lock_counter;
-		px4_indicate_external_reset_lockout(LockoutComponent::SystemShutdownLock, true);
-		return pthread_mutex_unlock(&shutdown_mutex);
-	}
+static struct work_s shutdown_work = {};
 
-	return ret;
-}
+#define SHUTDOWN_ARG_IN_PROGRESS (1<<0)
+#define SHUTDOWN_ARG_REBOOT (1<<1)
+#define SHUTDOWN_ARG_TO_BOOTLOADER (1<<2)
+#define SHUTDOWN_ARG_TO_ISP (1<<3)
+#define SHUTDOWN_ARG_BL_CONTINUE_BOOT (1<<4)
+
+static uint8_t shutdown_args = 0;
+
+class ShutdownAckCallback;
 
-int px4_shutdown_unlock()
+typedef struct {
+	ShutdownAckCallback *cb;
+	bool acked;
+} shutdown_ack_hook;
+
+static constexpr int max_shutdown_hooks = 10;
+static shutdown_ack_hook shutdown_hooks[max_shutdown_hooks] = {0};
+
+static hrt_abstime shutdown_time_us = 0;
+static constexpr hrt_abstime shutdown_timeout_us =
+	5_s; ///< force shutdown after this time if modules do not respond in time
+
+class ShutdownAckCallback : public uORB::SubscriptionCallback
 {
-	int ret = pthread_mutex_lock(&shutdown_mutex);
+public:
+	ShutdownAckCallback(uint8_t instance) :
+		uORB::SubscriptionCallback(ORB_ID(shutdown_ack), 0, instance)
+	{}
+
+	void call()
+	{
+		int ret = 0;
+		shutdown_ack_s ack_msg;
+
+		if (update(&ack_msg)) {
+
+			ret = pthread_mutex_trylock(&shutdown_hook_mutex);
```

Comment:
We shouldn't need a mutex here? Every call should be atomic by the uORB framework itself? Are you seeing problems with it?